### PR TITLE
Querystring operator for objectbrowser

### PR DESCRIPTION
**Pull request by cekk**
This is the backend part for https://github.com/plone/volto/pull/3342

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -258,8 +258,28 @@ existing ones. These are know to work well with Volto layout and grid system::
 sure your add-on's profiles are applied AFTER this one.**
 
 
+Queryparser operation for ObjectBrowser
+---------------------------------------
+
+There is a new operation "*plone.app.querystring.operation.string.objectBrowserReference*" for path index in querystring.
```
Comment (stevepiercy) [inline README.rst:264]:
```suggestion
There is a new operation `plone.app.querystring.operation.string.objectBrowserReference` for path index in `querystring`.
```

File: README.rst [hunk 2]
```diff
@@ -258,8 +258,28 @@ existing ones. These are know to work well with Volto layout and grid system::
...
+
+This is needed for pull-request `#3342 <https://github.com/plone/volto/pull/3342>`_  in Volto that enable the ObjectBrowser widget
```
Comment (stevepiercy) [inline README.rst:266]:
```suggestion
This is needed for pull request `#3342 <https://github.com/plone/volto/pull/3342>`_  in Volto that enables the `ObjectBrowser` widget
```

File: README.rst [hunk 3]
```diff
@@ -258,8 +258,28 @@ existing ones. These are know to work well with Volto layout and grid system::
...
+in listing blocks path criteria to select a path based on its UID and not a static path name.
+
+Expected data stored in "v" value of the criteria should be like this::
```
Comment (stevepiercy) [inline README.rst:269]:
```suggestion
Expected data stored in `v` value of the criteria should be similar to the following.

.. code-block:: python
```

File: README.rst [hunk 4]
```diff
@@ -258,8 +258,28 @@ existing ones. These are know to work well with Volto layout and grid system::
...
+
+  {
+      "object": {
+          "UID": "123456789",
+          "Title": "Foo",
+          "path": "/foo"
+      },
+      "depth": 1
+
```
Comment (stevepiercy) [inline README.rst:278]:
```suggestion
```

File: README.rst [hunk 5]
```diff
@@ -258,8 +258,28 @@ existing ones. These are know to work well with Volto layout and grid system::
...
+  }
+
```
Comment (stevepiercy) [inline README.rst:280]: Perhaps use 4-space indentation consistently above, including the initial indentation within `code-block` directive, assuming it is Python.

File: src/plone/volto/queryparser.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    We expect row.values to be something like this:
+
+    {
+        "object": {
+            "UID": "123456789",
+            "Title": "Foo",
+            "path": "/foo"
+        },
+        "depth": 1
+
```
Comment (stevepiercy) [inline src/plone/volto/queryparser.py:15]:
```suggestion
```

**Review by stevepiercy — CHANGES_REQUESTED** [review]
Some polish on the readme and docstrings.

File: README.rst (hunk 1 above)
Comment (jensens) [reply README.rst:264]: I would skip "new", since it is old/normal soon.

**Comment by jensens** [comment]
Why here? Don't we want this in Plone Classic UI collections?

**Comment by cekk** [comment]
@jensens because classic ui can't handle this operator.
There is a custom widget in mockup/whatever for path criteria and we should edit it or create a new one for this operator.

Classic ui works well with absolutePath and store also uids, but in volto we can't use it because we need more infos in the widget (see the discussion linked on my first comment).

We had two options: this pr, or stay in a "classic way" where Volto use absolutePath operator and store something like "the_uid_of_the_folder::1" and do an additional restapi call to get object infos for that UID to correctly draw the widget.

This was the easier one but i don't know if it's the right one for volto/classic integration

**Comment by jensens** [comment]
@cekk I would prefer to keep the querystrings compatible between classic ui and volto. Adding metadata to the value is most ugly. 

I am sure there is a better solution once we discuss this in a broader group on a sprint or in an dedicated meeting with more core dev folks from classic and volto. I propose to postpone this until there is a less hacky solution found.

**Comment by davisagli** [comment]
I tend to agree with @jensens. In essence this operation is about querying by container UID. It makes sense that the title and path may be needed for UI purposes, but they should not be persisted when the querystring is stored, because they could become out of date. Another possibility would be for plone.restapi to add the title and path when the querystring is serialized and remove them when it is deserialized. That way volto would have the necessary information without needing to make additional requests, but it would not be stored permanently.

**Comment by cekk** [comment]
Can we still register a new operator to use with path? It could work like classic ones on backend (store only a string), but we need it basically to tell Volto which widget to use.

With that custom widget, Volto could for example made a separate api call (only in edit mode) to get the full-data for the stored UID.

I don't see problems with this implementation with classic Plone because with plone.volto you can't use Collections unless you re-enable them manually (and you could ignore that new operator for path).

**Comment by cekk** [comment]
Ok, nevermind..we don't need this customization.
See related volto pr for updates.

**Pull request closed without merging**